spark/src/main/java/com/netflix/iceberg/spark/source/Reader.java

```diff
@@ -273,14 +282,23 @@ private Schema lazyExpectedSchema() {
     private Iterator<InternalRow> currentIterator = null;
     private Closeable currentCloseable = null;
     private InternalRow current = null;
+    private Iterator<InputFile> inputFiles = null;
```

Comment:
Maintaining two iterators separately leaves the risk of advancing one iterator and forgetting to advance the other iterator upon future modifications of this code. I wonder if we can build this in such a manner such that we can use a single iterator. For example we can make a struct containing both the `FileScanTask` and the `InputFile` associated with that task (that may have been decrypted), then make an iterator of that struct.

Reply:
Hmm ok I wasn't what the best way to do this was because I need to get all the `InputFiles` into a single iterator to pipe them through the `EncryptionManager`, but then I'd have to group them back with their respective FileScanTasks. A few options that I thought of are:

1. Create an Iterator that combines the two different iterators, ensuring that they're always used at the same time (see latest commit)
2. Create a map from the `FileScanTask` to the `InputFile` locations, as well as a map from the `InputFile` location to the decrypted `InputFile`, and use those maps to associate the appropriate decrypted `InputFile` with the `FileScanTask`... lots of bookkeeping though and a bit confusing. 
3. Don't use the encryption batch API. This makes it pretty straightforward to keep everything as a single iterator. 

Please let me know if I'm missing something though!

Reply:
We talked offline and decided to build a `Map<String, InputFile>` to map the paths to the input file instances (decrypted) that create the streams against that path. When we reach a `FileScanTask` we look up this map to find the `InputFile` that creates the associated stream. This assumes the invariant that `fileIo.newInputFile` and `encryptionManager#decrypt` will return paths that are consistent with their input parameters. We can check for this via `Preconditions`.